### PR TITLE
spotube: 3.8.3 → 3.9.0 | add aarch64-linux support.

### PR DESCRIPTION
**Pull request by damidoug**
## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/sp/spotube/package.nix [hunk 1]
```diff
@@ -79,8 +78,13 @@ let
     inherit pname version meta;
 
     src = fetchArtifact {
-      filename = "Spotube-linux-x86_64.deb";
-      hash = "sha256-x75ie9FXunClMT+YZVFlvl2VSDl933QYMRpEYjJ8YhY=";
+      filename =
```
Comment (drupol) [inline pkgs/by-name/sp/spotube/package.nix:81]: What you could also do is the following:

```nix
src = let
  sources = {
    "aarch64-linux" = {
      filename = "Spotube-linux-aarch64.deb";
      hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
    };
    "x86_64-linux" = {
      filename = "Spotube-linux-x86_64.deb";
      hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
    };
  };

  source = sources.${stdenv.hostPlatform.system};

in fetchArtifact {
  inherit (source) filename hash;
};
```

Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix:81]: I agree, that's a bit more readable.

An alternative would be something like this:
```nix
src = let
  sources = {
    "aarch64-linux" = fetchArtifact {
      filename = "Spotube-linux-aarch64.deb";
      hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
    };
    "x86_64-linux" = fetchArtifact {
      filename = "Spotube-linux-x86_64.deb";
      hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
    };
  };
in sources.${stdenv.hostPlatform.system};
```

**Review by TomaSajt — DISMISSED** [review]
`old -> new` should be used instead of `old > new` according to the [commit conventions](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#commit-conventions) for version updates.

Also, please put the maintainer list commit before the main commit.

*one way to do this: `git rebase -i HEAD~2` and swap the two commits*

File: pkgs/by-name/sp/spotube/package.nix (hunk 1 above)
Comment (drupol) [reply pkgs/by-name/sp/spotube/package.nix:81]: Indeed, even better :)

Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: Thanks for the feedback, I applied some changes in the file following this way.

File: pkgs/by-name/sp/spotube/package.nix [hunk 2]
```diff
@@ -1,63 +1,63 @@
...
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
+        "aarch64-darwin" = fetchArtifact {
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
+      };
+    in
+    sources.${stdenv.hostPlatform.system};
```
[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: Changed.

File: pkgs/by-name/sp/spotube/package.nix [hunk 3]
```diff
@@ -1,63 +1,63 @@
...
+          hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
+        };
+        "x86_64-darwin" = fetchArtifact {
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
+        "aarch64-darwin" = fetchArtifact {
+          filename = "Spotube-macos-universal.dmg";
+          hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
+        };
```
[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: Personally, I think it's better to have the sources attrset contain the derivations instead of just the data.

This way we could expose `sources` as `passthru.sources` which would allow us to easily fetch and update all FOD hashes at the same time.

Example from my recent pending PR:
https://github.com/NixOS/nixpkgs/blob/645e841fa609a8df25fc9e269280044a24bfc054/pkgs/by-name/vo/voicevox-core/package.nix#L13-L44

In this case, we would have to do `passthru = { inherit sources; };` because we're using two different builders

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: so is better keep on passthru like your code above or keep the way as my past commit?

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: If you don't have a preference, please change it to this one.

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: that's correct?

```
passthru.sources =
    let
      fetchArtifact =
        { filename, hash }:
        fetchurl {
          url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
          inherit hash;
        };
    in
    {
      "aarch64-linux" = fetchArtifact {
        filename = "Spotube-linux-aarch64.deb";
        hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
      };
      "x86_64-linux" = fetchArtifact {
        filename = "Spotube-linux-x86_64.deb";
        hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
      };
      "x86_64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
      "aarch64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
    };
  src = finalAttrs.passthru.sources.${stdenv.hostPlatform.system};
  
  ```

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: it should be something like this:
```nix
sources = ...
src = sources.${stdenv.hostPlatform.system};

linux = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
darwin = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]:
```
sources =
    let
      fetchArtifact =
        { filename, hash }:
        fetchurl {
          url = "https://github.com/KRTirtho/spotube/releases/download/v${version}/${filename}";
          inherit hash;
        };
    in
    {
      "aarch64-linux" = fetchArtifact {
        filename = "Spotube-linux-aarch64.deb";
        hash = "sha256-KBuUAgUU6c/rBkkbpYjSarzckIoi+gRtCkumvtaoras=";
      };
      "x86_64-linux" = fetchArtifact {
        filename = "Spotube-linux-x86_64.deb";
        hash = "sha256-vzzK3csyKYP6fKKIoysziBsc2tqrg5LXS/6KoXBtNVI=";
      };
      "x86_64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
      "aarch64-darwin" = fetchArtifact {
        filename = "Spotube-macos-universal.dmg";
        hash = "sha256-wwIIKY+bmMJZigc2AK/QMg142uvZ+D6LOddzedJM5f8=";
      };
    };

  src = sources.${stdenv.hostPlatform.system};
```

```
linux = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
darwin = stdenv.mkDerivation {
  ...
  passthru = { inherit sources; }
  ...
};
```

have some way to test this passthru before I push?

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: `nix-build -A spotube.sources` should build all the sources

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]:
```
[nix-shell:~/Downloads/nixpkgs]$ nix-build -A spotube.sources
these 2 derivations will be built:
  /nix/store/j4p94j0n8z6gj3wd17nk4fjf4i52hq3j-Spotube-linux-x86_64.deb.drv
  /nix/store/l24nksf52jb9m1fmffqfylhn061dk9lg-Spotube-linux-aarch64.deb.drv
building '/nix/store/l24nksf52jb9m1fmffqfylhn061dk9lg-Spotube-linux-aarch64.deb.drv'...
building '/nix/store/j4p94j0n8z6gj3wd17nk4fjf4i52hq3j-Spotube-linux-x86_64.deb.drv'...

trying https://github.com/KRTirtho/spotube/releases/download/v3.9.0/Spotube-linux-aarch64.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

trying https://github.com/KRTirtho/spotube/releases/download/v3.9.0/Spotube-linux-x86_64.deb
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100 18.9M  100 18.9M    0     0  12.4M      0  0:00:01  0:00:01 --:--:-- 30.9M
100 19.2M  100 19.2M    0     0  10.6M      0  0:00:01  0:00:01 --:--:-- 20.8M
/nix/store/h4cacazy1imwddy70khln5gah2kljzlz-Spotube-macos-universal.dmg
/nix/store/9sk5igi898aiz6jzx9585yc445a8rfys-Spotube-linux-aarch64.deb
/nix/store/h4cacazy1imwddy70khln5gah2kljzlz-Spotube-macos-universal.dmg
/nix/store/wgsnv01ixhq0yl05zj1722f28v60y0d8-Spotube-linux-x86_64.deb
```

cool? sorry for asking but I never had use passthru.

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: This is what was supposed to happen, so it's working.

`stdenv.mkDerivation { passthru.foo = "bar"; }` is esentially just `stdenv.mkDerivation { foo = "bar"; }`; but `foo` won't be converted into a bash variable. This is useful, because that conversion would evaluate the expression, which we might not want (in this case, we don't want all the sources to be passed to the builder).

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: Okay, thanks for the explication, I'll use more it on futures pkgs.

More one question, is really necessary pass inherit meta on linux and Darwin?

```
darwin = stdenv.mkDerivation {
    inherit
      pname
      version
      meta
      src
      ;
}

linux = stdenv.mkDerivation {
    inherit
      pname
      version
      meta
      src
      ;
 }
```

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: Yes, since it includes the description and other stuff needed.

[the remark this answers is not in the document]
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix]: Because we are not using meta.description example inside of linux or Darwin.

[the remark this answers is not in the document]
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix]: `meta` is just metadata, it's not supposed to be directly used by the package itself anywhere.
I hope that answers your question.

File: maintainers/maintainer-list.nix [hunk 4]
```diff
@@ -4911,6 +4911,12 @@
     github = "damhiya";
     githubId = 13533446;
   };
+  damidoug = {
+    email = "contact@damidoug.dev";
+    github = "damidoug";
+    githubId = 75175586;
+    name = "Douglas Damiano";
+  };
```
Comment (TomaSajt) [inline maintainers/maintainer-list.nix:4919]: You added yourself to the maintainer list, but did not add yourself as a maintainer to any package. Was this a mistake?

File: pkgs/by-name/sp/spotube/package.nix [hunk 5]
```diff
@@ -121,5 +126,26 @@ let
           --prefix PATH : ${lib.makeBinPath [ xdg-user-dirs ]}
     '';
   };
+
+  meta = with lib; {
```
Comment (TomaSajt) [inline pkgs/by-name/sp/spotube/package.nix:130]: Could you revert the changes made to `meta` (other than adding `aarch64-linux to platforms)?
`with lib;` is not very encouraged nowadays.

File: maintainers/maintainer-list.nix (hunk 4 above)
Comment (damidoug) [reply maintainers/maintainer-list.nix:4919]: I have an PR (#368618) to maintainer-list.nix but it was rejected, @drupol told me first to apply pr on pkgs.

File: pkgs/by-name/sp/spotube/package.nix (hunk 5 above)
Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix:130]:
```
meta = {
    description = "Open source, cross-platform Spotify client compatible across multiple platforms";
    longDescription = ''
      Spotube is an open source, cross-platform Spotify client compatible across
      multiple platforms utilizing Spotify's data API and YouTube (or Piped.video or JioSaavn)
      as an audio source, eliminating the need for Spotify Premium
    '';
    downloadPage = "https://github.com/KRTirtho/spotube/releases";
    homepage = "https://spotube.krtirtho.dev/";
    license = lib.licenses.bsdOriginal;
    mainProgram = "spotube";
    maintainers = with lib.maintainers; [ tomasajt ];
    platforms = [
      "x86_64-linux"
      "x86_64-darwin"
      "aarch64-darwin"
      "aarch64-linux"
    ];
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
  };
```

okay?

Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix:130]: This PR does not cleanly list package outputs after merging. do you know why this err?

File: pkgs/by-name/sp/spotube/package.nix [hunk 6]
```diff
@@ -1,63 +1,66 @@
...
+  makeBinaryWrapper,
+  makeWrapper,
   mpv-unwrapped,
-  xdg-user-dirs,
+  stdenv,
+  undmg,
   webkitgtk_4_1,
+  wrapGAppsHook3,
+  xdg-user-dirs,
 }:
```
Comment (TomaSajt) [inline pkgs/by-name/sp/spotube/package.nix:19]: I don't personally like alphabetical ordering. My own policy is this: `lib`, `stdenv`, and group the rest into categories. Those categories are then sorted alphabetically.
Could you undo this change.

In general, when making PRs, I try to change as few things as possible, so it's easier to review.

Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix:19]: okay sorry.

File: pkgs/by-name/sp/spotube/package.nix (hunk 5 above)
Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix:130]: > okay?

it should be the same as before this PR (except the platforms change)


> This PR does not cleanly list package outputs after merging. do you know why this err?

Your first commit title is not following the commit convention, and ofborg can't parse that.

Also, please squash your `spotube:` related commits into one commit.
*One way to do it is using: `git rebase -i HEAD~4`*

Comment (damidoug) [reply pkgs/by-name/sp/spotube/package.nix:130]: I have changed direct on GitHub, im gonna do it by git now, the title can be spotube: 3.8.3 → 3.9.0 | add aarch64-linux support or other way?

Comment (TomaSajt) [reply pkgs/by-name/sp/spotube/package.nix:130]: > I have changed direct on GitHub, im gonna do it by git now, the title can be spotube: 3.8.3 → 3.9.0 | add aarch64-linux support?

I think it's fine. Though I am not 100% sure about using |
 I usually use ,

**Comment by TomaSajt** [comment]
Please drop the unrelated commits
(Remove the lines when doing rebase -i)

**Comment by damidoug** [comment]
@TomaSajt you're saying about these commits from others in others files? I don't know why is showing it, when I use git rebase -I just have one line uncommented wrote "noop".

```
noop

# Rebase d836dc6ebf98..d836dc6ebf98 onto d836dc6ebf98 (1 command)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup [-C | -c] <commit> = like "squash" but keep only the previous
#                    commit's log message, unless -C is used, in which case
#                    keep only this commit's message; -c is same as -C but
#                    opens the editor
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
#         create a merge commit using the original merge commit's
#         message (or the oneline, if no original merge commit was
#         specified); use -c <commit> to reword the commit message
# u, update-ref <ref> = track a placeholder for the <ref> to be updated
#                       to this position in the new commits. The <ref> is
#                       updated at the end of the rebase
#
# These lines can be re-ordered; they are executed from top to bottom.
#
# If you remove a line here THAT COMMIT WILL BE LOST.
#
# However, if you remove everything, the rebase will be aborted.
#
~                                                              
```

I'm starting to use git and GitHub for PRs, so sorry if i'm doing something wrong.

**Comment by TomaSajt** [comment]
When doing `git rebase -i HEAD~5` the last 5 commits will be edited. In this case you edited the last 5 commits, but you only needed to edit the last 2, so the 3 extra commits got "edited" as well (even if nothing changed). You can remove those 3 commits by doing `git rebase -i HEAD~5`
This will appear:
```
pick 7b04c187df57 sfeed: 2.0 -> 2.1
pick e0d47e630c1d sfeed: modernize
pick 24f0766aca08 distribution: 3.0.0-rc.1 -> 3.0.0-rc.2
pick 418aa56d8859 maintainers: add damidoug
pick 6384c3751edb spotube: 3.8.3 → 3.9.0 | add aarch64-linux support
```

You will need to drop the 3 unneeded commits.

you either have to remove the unneeded commits:
```
pick 418aa56d8859 maintainers: add damidoug
pick 6384c3751edb spotube: 3.8.3 → 3.9.0 | add aarch64-linux support
```
or you have to write `drop` instead of `pick`
```
drop 7b04c187df57 sfeed: 2.0 -> 2.1
drop e0d47e630c1d sfeed: modernize
drop 24f0766aca08 distribution: 3.0.0-rc.1 -> 3.0.0-rc.2
pick 418aa56d8859 maintainers: add damidoug
pick 6384c3751edb spotube: 3.8.3 → 3.9.0 | add aarch64-linux support
```

**Review by TomaSajt — APPROVED** [review]
LGTM :)

File: pkgs/by-name/sp/spotube/package.nix (hunk 3 above)
[the remark this answers is not in the document]
Comment (drupol) [reply pkgs/by-name/sp/spotube/package.nix]: Thank you, TIL !

**Review by drupol — APPROVED** [review]
